### ISSUE TITLE
windows not in current desktop show as white rects

### ISSUE THREAD

**Issue by RaafatTurki** [dreamcat4/skippy-xd#20]
I'm on bspwm and after setting `showAllDesktops = true` all windows not in the current desktop show as white rects

![image](https://user-images.githubusercontent.com/16624558/177024975-bdc7c788-7e96-4761-8a55-d26869a6369e.png)

could be related to #17

**Comment by dreamcat4** [comment]
i see the same issue over here on KDE Plasma. Unfortunately I am not sure of the reason for this.

I am reading (just right now) that the other guy over in #17 was saying maybe if can add xcb calls (in addition to xlib).

And speaking in a broader sense... yes it would be nice to get this issue fixed. And (similar matter) if there was a reasonable way to get the previews for minimized windows too. (which IMHO could perhaps be just to keep around in cache the last captured view before minimization, and not discard it too soon, at least if we already are doing that).

Personally I don't have any objection to take such changes as pull requests. Just don't have the time to work on it myself these days.

So yes, if it requires adding xcb then there is a basic introduction to xcb here: https://www.x.org/wiki/guide/xlib-and-xcb/

BUT

I was also just looking in the code and stumbled across an interesting setting. Which I had not noticed before. The specific purpose of this setting is actually unclear to me, and also I didn't write this code, so it might be something else not-related. However if you take a look for; `avoidThumbnailsFromOtherScreens`, it appears the default is being set to `true` (if the setting was not specified / manually set in the config).

So maybe in the meantimes we could just try playing about with this and see if it's any help. Again, there's no guarantee this will do what we want it to. But might as well have a play. (and since the other proposal of implementing xcb calls would be far more involved than that).

The setting is labelled for `[general]` section of the ini config, so you would have to add it (if missing from your config) under that `[general]` tab, like this:

```ini
[general]
...
avoidThumbnailsFromOtherScreens = false
....


```

**Comment by dreamcat4** [comment]
actually... that did not work for me, over here on kde plasma. Didn't seem to do anything

**Comment by RaafatTurki** [comment]
Yeah that setting does nothing on my end as well, I unfortunately don't have the time to PR this as well, I'll see what I can do in the future.

**Comment by jzbor** [comment]
I think the issure might be related to [this line](https://github.com/dreamcat4/skippy-xd/blob/b2cf2e242315fcd6dc941c899a25cd080c0e45e4/src/clientwin.c#L167). I know bspwm and a lot of other WMs unmap windows when they are not on the current desktop. So their state is probably `IsUnmapped` and not `IsViewable`. So I think this is also related to #15. I don't know whether the composition extension allows for displaying unmapped windows though, so the fix might not be straight forward. I don't think xcb is required though Xlib can do almost everything xcb can although arguably with performance implications.

**Comment by RaafatTurki** [comment]
That's interesting.
It's worth mentioning this issue also happens on herbstluftwm.

**Comment by dreamcat4** [comment]
can read a little more about it here

https://xpert.xfree86.narkive.com/xvCWsd9O/about-map-state

if  the window has been minimized then it says it gets unmapped implicitly. so presumably that unmapping is occuring elsewhere in xwindows system... not sure myself how it could be possible to change that

requires further research / looking into. but good find in the skippy source code!

**Comment by dreamcat4** [comment]
or we could just change that line and see what happens... whether it crashes or not while subsequently trying to generate the thumbnail

**Comment by jzbor** [comment]
> or we could just change that line and see what happens... whether it crashes or not while subsequently trying to generate the thumbnail

I just tried that - I also uncommented all other lines referencing the mapped state. It behaved pretty much the same as before showing a mixture of white and blue rectangles.

The unmapping is done by the window manager as soon as you switch the desktop. I am not sure, but I think remapping it manually could confuse some window-managers into thinking the application wants to be displayed again.

[This is the relevant section in the ICCCM specification](https://x.org/releases/X11R7.6/doc/xorg-docs/specs/ICCCM/icccm.html#changing_window_state)

By the way dwm for example does not unmap windows, but instead moves them to an invisible part of the screen (x or y < 0), which is why this issue does not occur for dwm for example. Some other window managers might have other ways of hiding a window which do not unmap the window, but most window managers use this mechanism, as it is pretty simple to implement. 

From the client side this looks/behaves exactly as minimizing/iconifying it does, which is why skippy-xd also has issues with minimized clients.

**Comment by dreamcat4** [comment]
ah interesting...

so then i would suspect maybe there is another part of skippy code that is reacting to the implicit unmapping event. when a window is being iconified / minimized. and that other callback hook is then clearing the thumbnail image.

so maybe that would be a possible way to modify skippy, not to throw it away for implicit unmappings only. {however if the window is truly being destroyed then you would want to still be throwing it away)

that sort of a behaviour could then be optionally put under configuration, if not already existing an appropriate configuration flag for it already

**Comment by jzbor** [comment]
One other issue with X11 is that it does not store the contents of unmapped windows by default. This can be worked around with the composite extension, but in the case of calling skippy-xd not as a daemon it might still be very buggy, as the window contents may not be stored anywhere by the X11 server, so there might not be a possibility to draw those without redrawing/remapping the window first...

**Comment by RaafatTurki** [comment]
@jzbor tbh I didn't even know skippy-xd can be used without daemonizing. Such tools are always daemonized (think clipboard managers, window compositors, input layout switchers and even screenshooting programs).

**Comment by felixfung** [comment]
Fixed in #41
